Jenkinsfile

```diff
@@ -0,0 +1,158 @@
+        HOME = "${env.WORKSPACE}"
+        PATH = "${env.JAVA_HOME}/bin:${env.PATH}"
+        MAVEN_CONFIG = "${params.MAVEN_CONFIG} ${env.MAVEN_CONFIG}"
+      }
+      stages {
+        /**
+         Checkout the code and stash it, to use it on other stages.
+        */
+        stage('Checkout') {
+          steps {
+            deleteDir()
+            gitCheckout(basedir: "${BASE_DIR}", githubNotifyFirstTimeContributor: true)
+            stash allowEmpty: true, name: 'source', useDefaultExcludes: false
+          }
+        }
+        /**
+        Build on a linux environment.
+        */
+        stage('Build') {
+          options { skipDefaultCheckout() }
+          steps {
+            withGithubNotify(context: 'Build', tab: 'artifacts') {
+              deleteDir()
+              unstash 'source'
+              dir(BASE_DIR){
+                sh """#!/bin/bash
+                set -euxo pipefail
+                ./mvnw clean install -DskipTests=true -Dmaven.javadoc.skip=true
+                ./mvnw license:aggregate-third-party-report -Dlicense.excludedGroups=^co\\.elastic\\.
+                """
+                archiveArtifacts allowEmptyArchive: true, onlyIfSuccessful: true,
+                                 artifacts: 'target/site/aggregate-third-party-report.html'
+              }
+              stash allowEmpty: true, name: 'build', useDefaultExcludes: false
+            }
+          }
+        }
+      }
+    }
+    stage('Tests') {
+      environment {
+        MAVEN_CONFIG = "${params.MAVEN_CONFIG} ${env.MAVEN_CONFIG}"
+        HOME = "${env.WORKSPACE}"
+        PATH = "${env.JAVA_HOME}/bin:${env.PATH}"
+      }
+      failFast true
+      parallel {
+        /**
+          Run only unit test.
+        */
+        stage('Unit Tests') {
+          agent { label 'linux && immutable' }
+          options { skipDefaultCheckout() }
+          when {
+            beforeAgent true
+            expression { return params.test_ci }
+          }
+          steps {
+            withGithubNotify(context: 'Unit Tests', tab: 'tests') {
+              deleteDir()
+              unstash 'build'
+              dir(BASE_DIR){
+                sh './mvnw test'
+              }
+            }
+          }
+          post {
+            always {
+              junit(allowEmptyResults: true, keepLongStdio: true,
+                    testResults: "${BASE_DIR}/**/junit-*.xml,${BASE_DIR}/**/TEST-*.xml")
+            }
+          }
+        }
+        /**
+          Build javadoc files.
+        */
+        stage('Javadoc') {
+          agent { label 'linux && immutable' }
+          options { skipDefaultCheckout() }
+          when {
+            beforeAgent true
+            expression { return params.doc_ci }
+          }
+          steps {
+            withGithubNotify(context: 'Javadoc') {
+              deleteDir()
+              unstash 'build'
+              dir(BASE_DIR){
+                sh './mvnw compile javadoc:javadoc'
```

Comment:
What do we do with those javadocs? Are we pushing them to anywhere, or do we need to store them in Jenkins?

Reply:
That's a good point, this particular pipeline is a subset of the one for the apm-agent-java. As I didn't want to propose another workflow but the same as done with that particular java repo.

So, answering your question, the docs stage happens within the test stage in parallel as validation rather than archiving those files somewhere else.

I guess, the maven project should drive the release process and what artifacts should be released to the central maven repo.